packages/@lwc/engine-server/src/context.ts

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2023, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import {
+    createContextProviderWithRegister,
+    getAssociatedVMIfPresent,
+    LightningElement,
+    WireAdapterConstructor,
+    WireContextSubscriptionPayload,
+    WireContextSubscriptionCallback,
+} from '@lwc/engine-core';
+import { isUndefined } from '@lwc/shared';
+import {
+    HostElement,
+    HostNodeType,
+    HostTypeKey,
+    HostParentKey,
+    HostParentNode,
+    HostHostKey,
+    HostContextProvidersKey,
+} from './types';
+
+export function createContextProvider(adapter: WireAdapterConstructor) {
+    return createContextProviderWithRegister(adapter, registerContextProvider);
+}
+
+function registerContextProvider(
+    elm: HostElement | LightningElement,
+    adapterContextToken: string,
+    onContextSubscription: WireContextSubscriptionCallback
+) {
+    const vm = getAssociatedVMIfPresent(elm);
+    if (!isUndefined(vm)) {
+        elm = vm.elm;
+    }
+
+    const contextProviders = (elm as HostElement)[HostContextProvidersKey];
+    if (!contextProviders) {
```

Comment:
```suggestion
    if (!isUndefined(contextProviders)) {
```

for consistency with the rest of this file, yeah it's the server so it doesn't matter but we may as well